src/airtable.ts

```diff
@@ -1,77 +1,52 @@
-import AirtableSDK, { FieldSet } from "airtable"
-import { z } from "zod"
-import { SelectQueryParamsZ } from "./types/queryParams"
-import { FieldT } from "./types/airtableFields"
+// if the user wants a singleton to only have to get the api key once etc, then this returns curried functions
+// can also provide a default base id that is passed to all functions if a specific baseId is not provided
 
-/**
- * @class ZodAirTable
- */
-export default class ZodAirTable<
-	T extends z.ZodType<any, any, z.RecordType<string, FieldT>>
-> {
-	private table: AirtableSDK.Table<FieldSet>
-	private schema: T
-	private listSchema: z.ZodArray<
-		z.ZodObject<{ id: z.ZodString; createdTime: z.ZodString; fields: T }>
-	>
-	/**
-	 * @constructor
-	 * @param apiKey: Airtable Personal Access Token - https://airtable.com/create/tokens
-	 * @param baseId: Airtable base/appId
-	 * @param tableId: Table name or tableId
-	 * @param schema: zod object({}) schema for fields
-	 */
-	constructor(args: {
-		apiKey: string
-		baseId: string
-		tableId: string
-		schema: T
-	}) {
-		this.table = new AirtableSDK({ apiKey: args.apiKey })
-			.base(args.baseId)
-			.table(args.tableId)
-		this.schema = args.schema
-		this.listSchema = z.array(
-			z.object({
-				id: z.string(),
-				createdTime: z.string(),
-				fields: args.schema,
-			})
-		)
-	}
+import AirtableSDK from "airtable"
+
+import { getRecords, GetRecordsWithAirtable } from "./getRecords"
+import { SchemaZodT } from "./types/schema"
+
+interface ZodAirtable {
+	apiKey: string
+	defaultBaseId?: string
+}
+
+export default function zodAirtable ({
+	apiKey,
+	defaultBaseId
+}: ZodAirtable) {
+	const airtable = new AirtableSDK({ apiKey })
 
-	/**
-	 * @function getAllRecords returns all records in the table
-	 * @param query Optional query params inculding filterByFormula, maxRecords, etc
-	 * @returns Array of validated records
-	 */
-	public listAllRecords = z
-		.function()
-		.args(
-			z.object({
-				query: SelectQueryParamsZ,
-				fieldEnum: z.string().optional(),
+	// TODO: auto-curry in TS?
+	if (!defaultBaseId) {
+		return {
+			getRecords: <T extends SchemaZodT>({
+				tableId,
+				schema,
+				baseId,
+				query
+			}: Omit<GetRecordsWithAirtable<T>, 'airtable'>) => getRecords({
+				airtable,
+				tableId,
+				schema,
+				baseId,
+				query
 			})
```

Comment:
Yeah totally - part of what I wanted to experiment with a bit was whether it would work to just allow for both approaches (i.e. singleton and/or assorted functions)? I think it probably can, the typings get a bit intense but if it's in a library and done then we don't have to worry about it as a consumer.

I also was trying to account for that use case where you may have multiple bases in an app that you need to call on - I can't recall us ever having needed that, at least in an app I've worked on, but it's not too hard I think to imagine a scenario where we might. That's why `defaultBaseId` is there as an option for the singleton approach - trying to recognise that most of the time you probably will want that, but you aren't forced to establish multiple singletons if you have multiple bases. Not that there is anything super wrong with multiple singletons IMO, it's just a little contrary to the basic idea of a singleton, and just means the Airtable SDK gets fired up multiple times which tbh isn't a big deal I don't think.

I'm not sure about requiring the `schema` though - that would restrict your usage of the zod-airtable singleton to a specific schema shape right? Which from where I'm at sounds like saying we can only query a specific table that matches the given schema? It felt to me like you'd want to pass the schema you are parsing results against for each call, as each call may be against a different table (let alone a different base). And even if against the same table, you might have different data parsing requirements for different calls - i.e. one call I want to parse the "Products" table records thoroughly, but another I'm just interested in "draft" Products and not all the fields have to be complete.

Reply:
I have 2 observations on this:

1. Similar to having a Singleton per base, I'm happy with a Singleton per table. It shouldn't matter how many we instantiate through the factory function. If the issue is that we are instantiating to many instances of the Airtable SDK we should solve this with dependency injection in the Singleton constructor/init args.

```
const singletonTable1 = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID,
  schema: z.object({
    id: z.string(), 
    fields: z.object({fieldName: z.string()})
  })
})

const singletonTable2 = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID,
  schema: z.object({
    id: z.string(), 
    fields: z.object({fieldName: z.string()})
  })
})

const table1Data = singletonTable1.getRecords()
const table2Data = singletonTable2.getRecords()
```

2. The alternative would be to pass in a more complex `zod schema` that represented all the table in the base you wanted to interact with. We could then do some magic under the hood to enable something like this:

```
const singleton = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID,
  schema: z.object({
    table1: z.object({id: z.string(), fields: z.object({fieldName: z.string()})}),
    table2: z.object({id: z.string(), fields: z.object({fieldName: z.string()})})
  })
})

const table1Data = singleton.table1.getRecords()
const table2Data = singleton.table2.getRecords()

// instead of current dx in test

const singleton = zodAirtable({
  apiKey: API_KEY,
  defaultBaseId: BASE_ID
})

const table1Data = await singleton.getRecords({
  tableId: TABLE1_ID,
  schema: z.object({
    Name: z.string(),
  })
})

const table2Data = await singleton.getRecords({
  tableId: TABLE2_ID,
  schema: z.object({
    Name: z.string(),
  })
})
```

Noticing what I specifically don't like about not having the tableId and Schema in the singleton setup is I then have to still dependency inject on every function I use from it. Which is at odds with the point of the pattern IMO.

Reply:
@iainkirkpatrick I've pushed a refactored version along the lines of what I wrote above.

TLDR: its simpler code that is easier to reason with and is what I was aiming for for the Singleton approach. ie setup multiple instances of zodAirtable in a single codebase.

Reply:
@danalexilewis I've just pulled down the code to take a look - my initial thought is that the term "singleton" starts to lose a bit of meaning if we are intending that generally users should instantiate multiple "singletons", i.e. one per table they want to query.

To me this kind of just boils down to a question of where we are suggesting the table-specifying code should go. IMO it's going to be common to query multiple tables in an app that uses Airtable, so that leads me to think that specifying a table should be on a per-query basis. Querying multiple bases IMO is way less common, which is why included the `defaultBaseId` argument, but I still wanted it to be possible if you used zod-airtable without forcing the user to have to do multiple singletons.

If you require the user to pass the tableId for each query, bear in mind you can still get the kind of setup that I'm understanding you want - you just define a wrapper function in your code that closes over the zod-airtable function with the tableId already written:

```
function getRecordsFromPersonsTable ({
	apiKey,
	schema,
	query
}) {
	return getRecords({
		apiKey,
		tableId: "tbl_PERSONS",
		schema,
		query
	}) 
}
```
You can have something like that in a `utils/airtable` file etc and just import it where you need (and of course you can close over any of the other arguments too).

At the moment with the changes you've made, if you do want to query multiple tables in your app, you either have to use the `getRecords` function directly or you have to instantiate the whole zod-airtable multiple times right? I should be clear, I don't think that's the end of the world, it's not going to use up tons of memory or anything, just IMO it's an odd pattern.

I'm happy for now to just go with what works though (including just not using anything from this branch, totally fine!), AFRA is the biggest Airtable codebase we have so thinking of what suits there makes sense to me.